crates/database/src/onchain_broadcasted_orders.rs

```diff
@@ -0,0 +1,242 @@
+use super::events::EventIndex;
+use crate::{Address, OrderUid, PgTransaction};
+use sqlx::{Executor, PgConnection};
+
+#[derive(Clone, Debug, Default)]
+pub struct OnchainOrderPlacement {
+    pub order_uid: OrderUid,
+    pub sender: Address,
+}
+
+#[derive(Clone, Debug, Default, sqlx::FromRow, Eq, PartialEq)]
+pub struct OnchainOrderPlacementRow {
+    pub uid: OrderUid,
+    pub sender: Address,
+    pub is_reorged: bool,
+    pub block_number: i64,
+    pub log_index: i64,
+}
+
+pub async fn last_block(ex: &mut PgConnection) -> Result<i64, sqlx::Error> {
+    const QUERY: &str = "\
+                (SELECT COALESCE(MAX(block_number), 0) FROM onchain_placed_orders);";
+    sqlx::query_scalar(QUERY).fetch_one(ex).await
+}
+
+pub async fn mark_as_reorged(
+    ex: &mut PgTransaction<'_>,
+    mark_from_block_number: i64,
+) -> Result<(), sqlx::Error> {
+    const QUERY_ONCHAIN_ORDERS: &str =
+        "UPDATE onchain_placed_orders SET is_reorged = true WHERE block_number >= $1;";
+    ex.execute(sqlx::query(QUERY_ONCHAIN_ORDERS).bind(mark_from_block_number))
+        .await?;
+    Ok(())
+}
+
+pub async fn append(
+    ex: &mut PgTransaction<'_>,
+    events: &[(EventIndex, OnchainOrderPlacement)],
+) -> Result<(), sqlx::Error> {
+    for (index, event) in events {
+        insert_onchain_order(ex, index, event).await?;
+    }
+    Ok(())
+}
+
+async fn insert_onchain_order(
+    ex: &mut PgConnection,
+    index: &EventIndex,
+    event: &OnchainOrderPlacement,
+) -> Result<(), sqlx::Error> {
+    const QUERY: &str = "\
+        INSERT INTO onchain_placed_orders (uid, sender, is_reorged, block_number, log_index) VALUES ($1, $2, false, $3, $4) \
+        ON CONFLICT (uid) DO UPDATE SET is_reorged = false, sender = $2, block_number = $3, log_index = $4;";
```

Comment:
ubernit: Do you mind breaking up the long line? Rust-fmt has issues with these longer lines:
```suggestion
    const QUERY: &str = r#"
        INSERT INTO onchain_placed_orders
            (uid, sender, is_reorged, block_number, log_index)
        VALUES ($1, $2, false, $3, $4)
        ON CONFLICT (uid) DO UPDATE SET
            is_reorged = false, sender = $2, block_number = $3, log_index = $4;
    "#;
```